README.md
# HW2-Spy
Show curated information about the given Halo Wars 2 players, including a brief overview of their most recent matches' first 12 minutes, by accessing the public API available at https://developer.haloapi.com (a free subscription key is needed).

## Install
```
pipx ensurepath
pipx install hw2-spy
```
## Usage examples

### TUI Mode
```
hw2-spy --key your_key_here --red xandy92 --yellow american_sklz --tui
```
![](https://github.com/rtstrats/hw2-spy/blob/de2e7cfed826eb93bb4a7f4d2f5eb5c2f16db1c5/src/assets/images/hw2-spy-2vs2-red.png?raw=True)

```
hw2-spy --key your_key_here --blue edgreenall --cyan seanagone --green motoguzzi91 --tui
```
![](https://github.com/rtstrats/hw2-spy/blob/de2e7cfed826eb93bb4a7f4d2f5eb5c2f16db1c5/src/assets/images/hw2-spy-3vs3-blue.png?raw=True)

### JSON Mode
```
hw2-spy --key your_key_here --red btc_hosticide

{
    "data": {
        "red": {
            "gamertag": "btc hosticide",
            "xp": 1947672,
            "level": "99",
            "mmr1vs1": "2.09",
            "csr1vs1": "1 Onyx",
            "tier1vs1": "1",
            "designation1vs1": "Onyx",
            "mmr2vs2": "3.26",
            "csr2vs2": "1 Onyx",
            "tier2vs2": "1",
            "designation2vs2": "Onyx",
            "mmr3vs3": "2.91",
            "csr3vs3": "1 Onyx",
            "tier3vs3": "1",
            "designation3vs3": "Onyx",
            "matches": [
                {
                    "Population": [
                        0,
                        12,
                        39,
                        24,
                        65,
                        70,
                        56
                    ],
                    "Units": {
                        "Marine": 16,
                        "Cyclop": 7,
                        "Hornet": 10,
                        "Flame": 13,
                        "Nightingale": 5,
                        "FlameCyclop": 1,
                        "Flamehog": 3,
                        "Scorpion": 1
                    },
                    "Turrets": [
                        "05:30"
                    ],
                    "Bases": [
                        "04:34",
                        "10:44"
                    ],
                    "Minis": [
                        "01:35",
                        "01:50",
                        "09:21"
                    ],
                    "GameMode": 3,
                    "MatchType": 3,
                    "PlaylistId": "548d864e-8666-430e-9140-8dd2ad8fbfcd",
                    "Leader": "Kinsano",
                    "T2": "04:40",
                    "T3": "16:52",
                    "Duration": "21:41",
                    "Date": "2023-09-24 01:36:09",
                    "Result": "Defeat"
                },
                {
                    "Population": [
                        0,
                        15,
                        33,
                        63,
                        57
                    ],
                    "Units": {
                        "Marine": 17,
                        "Sniper": 6,
                        "Flame": 29,
                        "FlameCyclop": 1,
                        "Flamehog": 3
                    },
                    "Turrets": [
                        "05:13"
                    ],
                    "Bases": [],
                    "Minis": [
                        "01:43",
                        "02:02",
                        "05:28"
                    ],
                    "GameMode": 3,
                    "MatchType": 3,
                    "PlaylistId": "548d864e-8666-430e-9140-8dd2ad8fbfcd",
                    "Leader": "Kinsano",
                    "Duration": "08:29",
                    "Date": "2023-09-24 01:24:41",
                    "Result": "Victory"
                },
                {
                    "Population": [
                        0,
                        10,
                        18,
                        12,
                        40,
                        20,
                        24
                    ],
                    "Units": {
                        "Grunt": 12,
                        "Shipmaster": 1,
                        "Marauder": 13,
                        "Elite": 13,
                        "Locust": 1,
                        "Reaver": 1
                    },
                    "Turrets": [],
                    "Bases": [
                        "10:15"
                    ],
                    "Minis": [
                        "03:02",
                        "03:10",
                        "09:57",
                        "12:04"
                    ],
                    "GameMode": 3,
                    "MatchType": 3,
                    "PlaylistId": "548d864e-8666-430e-9140-8dd2ad8fbfcd",
                    "Leader": "Shipmaster",
                    "T2": "05:00",
                    "T3": "12:16",
                    "Duration": "14:47",
                    "Date": "2023-09-24 01:07:51",
                    "Result": "Defeat"
                }
            ]
        }
    },
    "status": "Success"
}
```

## TODO
- Finish unit translations
- TUI improvements
  - Tcss file cleanup
  - Use gutter between grid cells
  - Adjust yellow / orange colors for population display
  - Use async + workers for api calls and improve reactiveness
  - Switch method between 1vs1 2vs2 3vs3 from tui
  - Multiline (using rich?) for units as info may be trunked in 3vs3 depending on the screen
  - Fix color of text when in light mode using textual builtin vars 
- Web mode, generate html & copy to webserver
- OCR mode 



CHANGELOG.md
## [1.1.4] - 2023-10-09

### Bugfixes

- Readme and changelog format corrections.

## [1.1.3] - 2023-10-08

### Bugfixes

- Removed unneeded toml dependency.

## [1.1.2] - 2023-10-08

### New features

- Added install section and minor tweaks in README file.

### Bugfixes

- Ensure cache path is created if it does not exist.
- Version now extracted from config file.

## [1.1.1] - 2023-10-08

### Bugfixes

- Fixed scripts definition in pyproject,toml
- Rebuilt lockfile

## [1.1.0] - 2023-10-08

### New features

- Added Changelog
- Added script definition for direct execution
- Version now extracted from pyproject.toml

### Bugfixes

- Upgraded urllib3 to >=2.0.6 because CVE-2023-43804
- Disabled HTTP redirects using redirects=False because CVE-2023-43804
- Increased about screen width

## [1.0.0] - 2023-09-27
- Initial MVP release


hw2_spy/__main__.py
#!/usr/bin/env python

"""HW2_Spy main module."""
import argparse
import json
import logging
import sys
from typing import Any

from rich.traceback import install

from hw2_spy import hw2_spy_config, hw2_spy_data


def main() -> None:  # noqa: PLR0915, PLR0912, C901
    """Run the main entry point of the program.

    This function is responsible for initializing the application,
    processing command-line arguments, and orchestrating the program's
    execution.

    Returns
    -------
    None
        This function doesn't return any values.

    Examples
    --------
    To get stats for the red player in json format,
    ```
    python hw2_spy.py --red red_gamertag --json
    ```

    To get stats for the blue, cyan and green players using a tui,
    ```
    python hw2_spy.py --blue blue_gt --cyan cyan_gt --green green_gt --tui
    ```
    """
    # Install rich tracebacks
    install(show_locals=True)
    # logging conf
    logging.basicConfig(filename="hw2_spy.log", level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")
    # get the application version
    version = hw2_spy_data.get_version()
    # cmd parser
    parser = argparse.ArgumentParser(
        description="Halo Wars 2 Spy allows to fetch information about last matches of players."
    )
    # Define the mutually exclusive options
    group = parser.add_mutually_exclusive_group(required=False)
    group.add_argument("-r", "--red", nargs=1, help="Set the red player")
    group.add_argument("-b", "--blue", nargs=1, help="Set the blue player")
    # Define the other options
    parser.add_argument("-o", "--orange", nargs=1, help="Set the orange player")
    parser.add_argument("-y", "--yellow", nargs=1, help="Set the yellow player")
    parser.add_argument("-c", "--cyan", nargs=1, help="Set the cyan player")
    parser.add_argument("-g", "--green", nargs=1, help="Set the green player")
    parser.add_argument("-t", "--tui", action="store_true", help="Enable TUI mode")
    parser.add_argument("-w", "--web", action="store_true", help="Enable Web mode")
    parser.add_argument("-j", "--json", action="store_true", help="Enable JSON mode")
    parser.add_argument("-k", "--key", nargs=1, help="Specify an API key")
    usage = (
        f"""HW2-Spy v.{version} \n"""
        """hw2-spy [-h] (-r RED [-y YELLOW [-o ORANGE] ] | -b BLUE """
        """[-c CYAN [-g GREEN]]) [-k KEY] [--tui] [--web] [--json]"""
    )
    parser.usage = usage
    args = parser.parse_args()
    # Validate the complex option relationships
    if args.green and not (args.cyan and args.blue):
        parser.error("-g option requires -b and -c")
    if args.cyan and not args.blue:
        parser.error("-c option requires -b")
    if args.orange and not (args.yellow and args.red):
        parser.error("-o option requires -r and -y")
    if args.yellow and not args.red:
        parser.error("-y option requires -r")
    # web and json options require gamertags passed via cmd
    if args.web and not (args.blue or args.red):
        parser.error("-web option requires at least -r or -b")
    if args.json and not (args.blue or args.red):
        parser.error("-json option requires at least -r or -b")
    # set the api key (needed in order to fetch data)
    api_key = None
    if args.key:
        api_key = str(args.key[0])
    elif hasattr(hw2_spy_config, "api_key"):
        api_key = hw2_spy_config.api_key
    else:
        output = {"Status": "Error: Api key not found."}
        print(json.dumps(output, indent=4))  # noqa: T201
        logging.exception("Api key not found. Quitting...")
        sys.exit(1)
    # Determine the mode by the given number of players
    if args.green or args.orange:
        mode = "3vs3"
    elif args.cyan or args.yellow:
        mode = "2vs2"
    elif args.red or args.blue:
        mode = "1vs1"
    else:
        # default when no gamertags given
        mode = "1vs1"
    #
    p1g = None
    p2g = None
    p3g = None
    if args.tui:
        from hw2_spy import hw2_spy_tui

        color = "blue"  # default
        if args.blue:
            color = "blue"
            p1g = args.blue[0]
            if args.cyan:
                p2g = args.cyan[0]
                if args.green:
                    p3g = args.green[0]
        if args.red:
            color = "red"
            p1g = args.red[0]
            if args.yellow:
                p2g = args.yellow[0]
                if args.orange:
                    p3g = args.orange[0]
        app = hw2_spy_tui.HW2SpyApp(mode, color, p1g, p2g, p3g, api_key)
        app.run()
    else:
        # set the HW2Api instance in order to centralise the access to the API
        hw2api = hw2_spy_data.HW2Api(key=api_key)
        # Clear the cache
        hw2api.clear_cache()
        # Get the stats for every specified player
        stats: dict[str, Any] = {}
        stats["data"] = {}
        if args.red:
            player = hw2_spy_data.PlayerStats(args.red[0], mode, hw2api)
            stats["data"]["red"] = player.export_json()
        if args.blue:
            player = hw2_spy_data.PlayerStats(args.blue[0], mode, hw2api)
            stats["data"]["blue"] = player.export_json()
        if args.yellow:
            player = hw2_spy_data.PlayerStats(args.yellow[0], mode, hw2api)
            stats["data"]["yellow"] = player.export_json()
        if args.cyan:
            player = hw2_spy_data.PlayerStats(args.cyan[0], mode, hw2api)
            stats["data"]["cyan"] = player.export_json()
        if args.orange:
            player = hw2_spy_data.PlayerStats(args.orange[0], mode, hw2api)
            stats["data"]["orange"] = player.export_json()
        if args.green:
            player = hw2_spy_data.PlayerStats(args.green[0], mode, hw2api)
            stats["data"]["green"] = player.export_json()
        stats["status"] = "Success"
        json_data = json.dumps(stats, indent=4)
        print(json_data)  # noqa: T201


if __name__ == "__main__":
    main()


hw2_spy/samples/get_samples.sh
#!/bin/sh
KEY="place_your_api_key_here"
GAMERTAG="L1am%20Wh1te"
GAMERTAG2="xandy92"
MATCH="bc300422-8c60-45b2-8784-5356f7d235c0"
# metadata
/usr/bin/curl "https://www.haloapi.com/metadata/hw2/csr-designations" -H "Ocp-Apim-Subscription-Key: $KEY" >> csr_designations.json
/usr/bin/curl "https://www.haloapi.com/metadata/hw2/leaders" -H "Ocp-Apim-Subscription-Key: $KEY" >> leaders.json
/usr/bin/curl "https://www.haloapi.com/metadata/hw2/playlists" -H "Ocp-Apim-Subscription-Key: $KEY" >> playlists.json
/usr/bin/curl "https://www.haloapi.com/metadata/hw2/spartan-ranks" -H "Ocp-Apim-Subscription-Key: $KEY" >> spartan-ranks.json
/usr/bin/curl "https://www.haloapi.com/metadata/hw2/game-objects" -H "Ocp-Apim-Subscription-Key: $KEY" >> game-objects.json
# stats
/usr/bin/curl "https://www.haloapi.com/stats/hw2/players/$GAMERTAG/matches?matchType=matchmaking" -H "Ocp-Apim-Subscription-Key: $KEY" >> player_match_history.json
/usr/bin/curl "https://www.haloapi.com/stats/hw2/matches/$MATCH" -H "Ocp-Apim-Subscription-Key: $KEY" >> match_result.json
/usr/bin/curl "https://www.haloapi.com/stats/hw2/matches/$MATCH/events" -H "Ocp-Apim-Subscription-Key: $KEY" >> match_events.json
/usr/bin/curl "https://www.haloapi.com/stats/hw2/playlist/379f9ee5-92ec-45d9-b5e5-9f30236cab00/rating?players=$GAMERTAG" -H "Ocp-Apim-Subscription-Key: $KEY" >> player_playlist_ratings.json
/usr/bin/curl "https://www.haloapi.com/stats/hw2/players/$GAMERTAG/stats/seasons/current" -H "Ocp-Apim-Subscription-Key: $KEY" >> player_season_stats_summary.json
/usr/bin/curl "https://www.haloapi.com/stats/hw2/players/$GAMERTAG/stats" -H "Ocp-Apim-Subscription-Key: $KEY" >> player_stats_summary.json
/usr/bin/curl "https://www.haloapi.com/stats/hw2/xp?players=$GAMERTAG,$GAMERTAG2" -H "Ocp-Apim-Subscription-Key: $KEY" >> player_xps.json


hw2_spy/hw2_spy_config.py
"""HW2_Spy configuration file."""
__version__ = "1.1.4"
# api_key: str = "your_key_here"  # noqa: ERA001
api_max_requests: int = 10
api_interval_seconds: int = 10

leaders: dict[int, str] = {
    1: "Cutter",
    2: "Isabel",
    3: "Anders",
    4: "Decimus",
    5: "Atriox",
    6: "Shipmaster",
    7: "Forge",
    8: "Kinsano",
    9: "Jerome",
    10: "Arbiter",
    11: "Johnson",
    12: "Colony",
    13: "Serina",
    14: "YapYap",
    15: "Pavium",
    16: "Voridus",
}

units: dict[str, str] = {
    "unsc_veh_warthog_01": "Warthog",
    "unsc_inf_generic_marine": "Marine",
    "unsc_veh_forgehog_01": "Forge",
    "unsc_inf_cyclops_01": "Cyclop",
    "unsc_veh_foxcannon_01": "unsc_veh_foxcannon_01",
    "cov_inf_generic_grunt": "Grunt",
    "cov_inf_lekgologoliath_01": "Goliath",
    "dlc3_pack2_units_covenant_structure_lekgolowall": "dlc3_pack2_units_covenant_structure_lekgolowall",
    "cov_inf_impervioushunter_01": "Colony",
    "cov_inf_jackal_01": "Elite",
    "cov_veh_skitterer_01": "Skitter",
    "cov_air_banshee_01": "Banshee",
    "cov_inf_generic_brutejumppack": "Brute",
    "cov_veh_bruteChopper_01": "Chopper",
    "unsc_inf_flameMarine_01": "Flame",
    "unsc_inf_johnson_hero_01": "Johnson",
    "unsc_bldg_johnsonbunker_01_mp": "Bunker",
    "unsc_inf_sniper_01": "Sniper",
    "unsc_veh_johnson_mantis_01": "Mantis",
    "unsc_bldg_siegedropTurret_01": "unsc_bldg_siegedropTurret_01",
    "fx_mine_lotus_01_mp": "fx_mine_lotus_01_mp",
    "unsc_air_hornet_01": "Hornet",
    "unsc_air_nightingale_01": "Nightingale",
    "cov_inf_gruntswarm_01": "Fodder",
    "fx_mine_ambushmine_01": "fx_mine_ambushmine_01",
    "cov_inf_gruntswarm_01_frommine": "cov_inf_gruntswarm_01_frommine",
    "cov_inf_generic_heavygrunt": "HeavyGrunt",
    "fx_mine_ambushmine_02": "fx_mine_ambushmine_02",
    "cov_inf_gruntgoblin01": "Goblin",
    "cov_inf_hunter_01": "Hunter",
    "cov_bldg_grunt_shadeturret_01": "cov_bldg_grunt_shadeturret_01",
    "cov_veh_locust_01": "Locust",
    "cov_bldg_grunt_shieldtower_01": "cov_bldg_grunt_shieldtower_01",
    "cov_inf_mortarwarlord_01": "Pavium",
    "cov_veh_prowler_01": "Marauder",
    "cov_veh_gorgon_01": "Reaver",
    "cov_inf_atrioxchosen_01": "Atriox",
    "cov_inf_engineer_01": "Engineer",
    "cov_veh_scarab_01": "Scarab",
    "unsc_veh_wolverine_01": "Wolverine",
    "fx_mine_plasma_01_mp": "fx_mine_plasma_01_mp",
    "cov_inf_generic_suicideGrunt": "SuicideGrunt",
    "fx_mine_rcontrolmine_01": "fx_mine_rcontrolmine_01",
    "unsc_inf_spartan_mpjerome_01": "Jerome",
    "unsc_inf_odst_01": "ODST",
    "unsc_air_vulture_01": "Vulture",
    "cov_inf_brutegrenadier_01": "Grenadier",
    "unsc_inf_spartan_mpdoug_01": "Douglas",
    "unsc_veh_scorpion_01": "Scorpion",
    "cov_inf_generic_arbitergruntsquad": "Grunt",
    "cov_inf_generic_arbiterenforcer": "Enforcer",
    "fx_mine_arbiter_stasismine_01": "fx_mine_arbiter_stasismine_01",
    "cov_veh_ghost_01": "Ghost",
    "cov_inf_eliteCommando_01": "Shipmaster",
    "cov_air_shroud_01": "Shroud",
    "unsc_air_serina_pegasus_01": "Frostraven",
    "unsc_bldg_serina_iceblock_01": "unsc_bldg_serina_iceblock_01",
    "fx_mine_unbrk_mine_01_mp": "fx_mine_unbrk_mine_01_mp",
    "cov_air_blisterback_01": "Blisterback",
    "cov_inf_bruteHonour_01": "cov_inf_bruteHonour_01",
    "unsc_inf_flamecyclops_01": "FlameCyclop",
    "unsc_veh_flamehog_01": "Flamehog",
}


ignored_units: list[str] = [
    "pow_gp_scatterbombDummy_01",
    "dlc3_pack2_units_covenant_structure_lekgolowall",
    "unsc_bldg_siegedropTurret_01",
    "unsc_bldg_johnsonbunker_01_mp",
    "fx_mine_lotus_01_mp",
    "fx_mine_ambushmine_01",
    "fx_mine_ambushmine_02",
    "cov_bldg_grunt_shadeturret_01",
    "cov_bldg_grunt_shieldtower_01",
    "fx_mine_plasma_01_mp",
    "fx_mine_rcontrolmine_01",
    "fx_mine_unbrk_mine_01_mp",
    "unsc_bldg_serina_iceblock_01",
    "fx_mine_arbiter_stasismine_01",
]

designations: dict[str, str] = {
    "1": "Bronze",
    "2": "Silver",
    "3": "Gold",
    "4": "Platinum",
    "5": "Diamond",
    "6": "Onyx",
    "7": "Champ",
}

results: dict[int, str] = {0: "Unknown", 1: "Victory", 2: "Defeat", 3: "Tie"}

play_lists: dict[str, list[str]] = {
    "1vs1": ["548d864e-8666-430e-9140-8dd2ad8fbfcd"],
    "2vs2": ["379f9ee5-92ec-45d9-b5e5-9f30236cab00"],
    "3vs3": ["4a2cedcc-9098-4728-886f-60649896278d"],
}

levels: dict[int, int] = {
    1: 0,
    2: 1400,
    3: 3000,
    4: 5000,
    5: 7500,
    6: 10400,
    7: 13500,
    8: 16800,
    9: 20300,
    10: 24000,
    11: 27700,
    12: 31500,
    13: 35400,
    14: 39400,
    15: 43500,
    16: 47700,
    17: 52100,
    18: 56600,
    19: 61200,
    20: 66000,
    21: 71000,
    22: 76300,
    23: 81800,
    24: 87500,
    25: 93400,
    26: 99500,
    27: 105800,
    28: 112300,
    29: 119000,
    30: 126000,
    31: 133000,
    32: 140200,
    33: 147600,
    34: 155200,
    35: 163000,
    36: 171000,
    37: 179100,
    38: 187300,
    39: 195600,
    40: 204000,
    41: 212600,
    42: 221400,
    43: 230400,
    44: 239600,
    45: 249000,
    46: 258600,
    47: 268400,
    48: 278600,
    49: 289200,
    50: 320000,
    51: 331000,
    52: 342100,
    53: 353300,
    54: 364600,
    55: 376000,
    56: 387500,
    57: 399100,
    58: 410800,
    59: 422600,
    60: 434500,
    61: 446600,
    62: 458900,
    63: 471400,
    64: 484100,
    65: 497000,
    66: 510100,
    67: 523400,
    68: 536900,
    69: 550600,
    70: 564500,
    71: 578700,
    72: 593200,
    73: 608000,
    74: 623100,
    75: 638500,
    76: 654200,
    77: 670200,
    78: 686500,
    79: 703100,
    80: 720000,
    81: 737300,
    82: 755000,
    83: 773100,
    84: 791600,
    85: 810500,
    86: 829800,
    87: 849500,
    88: 869600,
    89: 890100,
    90: 911000,
    91: 932400,
    92: 954300,
    93: 976700,
    94: 999600,
    95: 1023000,
    96: 1046900,
    97: 1071300,
    98: 1096200,
    99: 1170000,
}

# buildings = {"x": "y"}  # noqa: ERA001
